diff --git a/src/main/java/mx/unam/ciencias/icc/Estudiante.java b/src/main/java/mx/unam/ciencias/icc/Estudiante.java
index 2e6fcc0..4d18935 100644
--- a/src/main/java/mx/unam/ciencias/icc/Estudiante.java
+++ b/src/main/java/mx/unam/ciencias/icc/Estudiante.java
@@ -26,6 +26,10 @@ public class Estudiante {
                       int    cuenta,
                       double promedio,
                       int    edad) {
+        this.nombre    = nombre;
+        this.cuenta    = cuenta;
+        this.promedio  = promedio;
+        this.edad      = edad;
         // Aquí va su código.
     }
 
@@ -34,6 +38,7 @@ public class Estudiante {
      * @return el nombre del estudiante.
      */
     public String getNombre() {
+        return nombre;
         // Aquí va su código.
     }
 
@@ -42,6 +47,7 @@ public class Estudiante {
      * @param nombre el nuevo nombre del estudiante.
      */
     public void setNombre(String nombre) {
+        this.nombre = nombre;
         // Aquí va su código.
     }
 
@@ -50,6 +56,7 @@ public class Estudiante {
      * @return el número de cuenta del estudiante.
      */
     public int getCuenta() {
+        return cuenta;
         // Aquí va su código.
     }
 
@@ -58,6 +65,7 @@ public class Estudiante {
      * @param cuenta el nuevo número de cuenta del estudiante.
      */
     public void setCuenta(int cuenta) {
+        this.cuenta = cuenta;
         // Aquí va su código.
     }
 
@@ -66,6 +74,7 @@ public class Estudiante {
      * @return el promedio del estudiante.
      */
     public double getPromedio() {
+        return promedio;
         // Aquí va su código.
     }
 
@@ -74,6 +83,7 @@ public class Estudiante {
      * @param promedio el nuevo promedio del estudiante.
      */
     public void setPromedio(double promedio) {
+        this.promedio = promedio;
         // Aquí va su código.
     }
 
@@ -82,6 +92,7 @@ public class Estudiante {
      * @return la edad del estudiante.
      */
     public int getEdad() {
+        return edad;
         // Aquí va su código.
     }
 
@@ -90,6 +101,7 @@ public class Estudiante {
      * @param edad la nueva edad del estudiante.
      */
     public void setEdad(int edad) {
+        this.edad = edad;
         // Aquí va su código.
     }
 
@@ -98,6 +110,12 @@ public class Estudiante {
      * @return una representación en cadena del estudiante.
      */
     @Override public String toString() {
+        String datitos = String.format("Nombre   : %s\n" +
+                                       "Cuenta   : %09d\n" +
+                                       "Promedio : %2.2f\n" +
+                                       "Edad     : %d",
+                                       nombre, cuenta, promedio, edad);
+	    return datitos;
         // Aquí va su código.
     }
 
@@ -113,6 +131,16 @@ public class Estudiante {
         if (!(objeto instanceof Estudiante))
             return false;
         Estudiante estudiante = (Estudiante)objeto;
+        if ((this.nombre.equals(estudiante.nombre)) == false){
+            return false;
+        } else if ((this.cuenta == estudiante.cuenta) == false) {
+            return false;
+        } else if ((this.promedio == estudiante.promedio) == false){
+            return false;
+        } else if ((this.edad == estudiante.edad) == false){
+            return false;
+        }
+        return true;
         // Aquí va su código.
     }
 }
diff --git a/src/main/java/mx/unam/ciencias/icc/Lista.java b/src/main/java/mx/unam/ciencias/icc/Lista.java
index 106877b..069a4dd 100644
--- a/src/main/java/mx/unam/ciencias/icc/Lista.java
+++ b/src/main/java/mx/unam/ciencias/icc/Lista.java
@@ -26,6 +26,7 @@ public class Lista {
 
         /* Construye un nodo con un elemento. */
         private Nodo(Object elemento) {
+            this.elemento = elemento;
             // Aquí va su código.
         }
 
@@ -34,6 +35,7 @@ public class Lista {
          * @return el nodo anterior del nodo.
          */
         public Nodo getAnterior() {
+            return anterior;
             // Aquí va su código.
         }
 
@@ -42,6 +44,7 @@ public class Lista {
          * @return el nodo siguiente del nodo.
          */
         public Nodo getSiguiente() {
+            return siguiente;
             // Aquí va su código.
         }
 
@@ -50,6 +53,7 @@ public class Lista {
          * @return el elemento del nodo.
          */
         public Object get() {
+            return elemento;
             // Aquí va su código.
         }
     }
@@ -66,6 +70,7 @@ public class Lista {
      * @return la longitud de la lista, el número de elementos que contiene.
      */
     public int getLongitud() {
+        return longitud;
         // Aquí va su código.
     }
 
@@ -75,6 +80,7 @@ public class Lista {
      *         otro caso.
      */
     public boolean esVacia() {
+        return cabeza == null;
         // Aquí va su código.
     }
 
@@ -85,6 +91,18 @@ public class Lista {
      *                 si es distinto de <code>null</code>.
      */
     public void agregaFinal(Object elemento) {
+        if (elemento == null) {
+            return;
+        }
+        Nodo nuevito = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevito;
+        } else {
+            rabo.siguiente = nuevito;
+            nuevito.anterior = rabo;
+            rabo = nuevito;
+        }
+        longitud++;
         // Aquí va su código.
     }
 
@@ -95,9 +113,85 @@ public class Lista {
      *                 si es distinto de <code>null</code>.
      */
     public void agregaInicio(Object elemento) {
+        if (elemento == null) {
+            return;
+        }
+        Nodo nuevito = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevito;
+        } else {
+            cabeza.anterior = nuevito;
+            nuevito.siguiente = cabeza;
+            cabeza = nuevito;
+        }
+        longitud++;
         // Aquí va su código.
     }
 
+    /*
+    Metodo auxiliar para buscar el nodo donde esta un estudiante (elemento)
+    Obtener Nodo dando Estudiante
+    */
+    private Nodo buscaNodo(Object buscado) {
+        if (buscado == null) {
+            return null;
+        }
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento.equals(buscado)) {
+                return explorador;
+            }
+            explorador = explorador.siguiente;
+        }
+        return null;
+    }
+
+    /*
+    Metodo auxiliar para buscar el nodo donde esta un estudiante (elemento)
+    Obtener Nodo dando Estudiante
+    */
+    private Nodo buscaNodoConI(int i) {
+        if (i < 0 || i >= longitud) {
+            return null;
+        }
+        Nodo explorador = cabeza;
+        int contando = 0;
+        while (explorador != null) {
+            if (contando == i) {
+                return explorador;
+            }
+            explorador = explorador.siguiente;
+            contando++;
+        }
+        return null;
+    }
+
+    /*
+    Metodo auxiliar para eliminar un nodo
+    Eliminar Nodo dando el Nodo
+    */
+    private void desapareceNodo(Nodo condenado) {
+        if (condenado == null) {
+            return;
+        }
+        if (cabeza == rabo && cabeza == null) {
+            return;
+        }
+        longitud--;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else if (condenado == cabeza){
+            cabeza.siguiente.anterior = null;
+            cabeza = cabeza.siguiente;
+        } else if (condenado == rabo) {
+            rabo.anterior.siguiente = null;
+            rabo = rabo.anterior;
+        } else {
+            condenado.anterior.siguiente = condenado.siguiente;
+            condenado.siguiente.anterior = condenado.anterior;
+        }
+    }
+
     /**
      * Inserta un elemento en un índice explícito.
      *
@@ -113,6 +207,22 @@ public class Lista {
      *                 si es distinto de <code>null</code>.
      */
     public void inserta(int i, Object elemento) {
+        if (elemento == null) {
+            return;
+        }
+        if (i <= 0) {
+            agregaInicio(elemento);
+        } else if (i >= longitud) {
+            agregaFinal(elemento);
+        } else {
+            Nodo intercambio = buscaNodoConI(i);
+            Nodo nuevo = new Nodo(elemento);
+            intercambio.anterior.siguiente = nuevo;
+            nuevo.anterior = intercambio.anterior;
+            intercambio.anterior = nuevo;
+            nuevo.siguiente = intercambio;
+            longitud++;
+        }
         // Aquí va su código.
     }
 
@@ -122,6 +232,14 @@ public class Lista {
      * @param elemento el elemento a eliminar.
      */
     public void elimina(Object elemento) {
+        if (elemento == null) {
+            return;
+        }
+        Nodo marcado = buscaNodo(elemento);
+        if (marcado == null) {
+            return;
+        }
+        desapareceNodo(marcado);
         // Aquí va su código.
     }
 
@@ -131,6 +249,12 @@ public class Lista {
      *         <code>null</code> si la lista es vacía.
      */
     public Object eliminaPrimero() {
+        if (esVacia()) {
+            return null;
+        }
+        Object porEliminar = cabeza.elemento;
+        desapareceNodo(cabeza);
+        return porEliminar;
         // Aquí va su código.
     }
 
@@ -140,6 +264,12 @@ public class Lista {
      *         <code>null</code> si la lista es vacía.
      */
     public Object eliminaUltimo() {
+        if (esVacia()) {
+            return null;
+        }
+        Object porEliminar = rabo.elemento;
+        desapareceNodo(rabo);
+        return porEliminar;
         // Aquí va su código.
     }
 
@@ -150,6 +280,7 @@ public class Lista {
      *         <code>false</code> en otro caso.
      */
     public boolean contiene(Object elemento) {
+        return buscaNodo(elemento) != null;
         // Aquí va su código.
     }
 
@@ -158,6 +289,13 @@ public class Lista {
      * @return una nueva lista que es la reversa la que manda llamar el método.
      */
     public Lista reversa() {
+        Lista clonada = new Lista();
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            clonada.agregaInicio(explorador.elemento);
+            explorador = explorador.siguiente;
+        }
+        return clonada;
         // Aquí va su código.
     }
 
@@ -167,6 +305,13 @@ public class Lista {
      * @return una copiad de la lista.
      */
     public Lista copia() {
+        Lista clonada = new Lista();
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            clonada.agregaFinal(explorador.elemento);
+            explorador = explorador.siguiente;
+        }
+        return clonada;
         // Aquí va su código.
     }
 
@@ -174,6 +319,8 @@ public class Lista {
      * Limpia la lista de elementos, dejándola vacía.
      */
     public void limpia() {
+        cabeza = rabo = null;
+        longitud = 0;
         // Aquí va su código.
     }
 
@@ -183,6 +330,11 @@ public class Lista {
      *         es vacía.
      */
     public Object getPrimero() {
+        if (esVacia()) {
+            return null;
+        } else {
+            return cabeza.elemento;
+        }
         // Aquí va su código.
     }
 
@@ -192,6 +344,11 @@ public class Lista {
      *         es vacía.
      */
     public Object getUltimo() {
+        if (esVacia()) {
+            return null;
+        } else {
+            return rabo.elemento;
+        }
         // Aquí va su código.
     }
 
@@ -203,6 +360,19 @@ public class Lista {
      *         elementos en la lista.
      */
     public Object get(int i) {
+        if (i < 0 || i >= longitud) {
+            return null;
+        }
+        Nodo explorador = cabeza;
+        int contando = 0;
+        while (explorador != null) {
+            if (contando == i) {
+                return explorador.elemento;
+            }
+            explorador = explorador.siguiente;
+            contando++;
+        }
+        return null;
         // Aquí va su código.
     }
 
@@ -213,6 +383,16 @@ public class Lista {
      *         no está contenido en la lista.
      */
     public int indiceDe(Object elemento) {
+        int medidor = 0;
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento.equals(elemento)) {
+                return medidor;
+            }
+            explorador = explorador.siguiente;
+            medidor++;
+        }
+        return -1;
         // Aquí va su código.
     }
 
@@ -221,6 +401,23 @@ public class Lista {
      * @return una representación en cadena de la lista.
      */
     @Override public String toString() {
+        if (esVacia()) {
+            String cadenitaRegreso = "[]";
+            return cadenitaRegreso;
+        }
+        String cadenitaRegreso = "[";
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento != null) {
+                cadenitaRegreso += explorador.elemento;
+            }
+            if (explorador.siguiente != null) {
+                cadenitaRegreso += ", ";
+            }
+            explorador = explorador.siguiente;
+        }
+        cadenitaRegreso += "]";
+        return cadenitaRegreso;
         // Aquí va su código.
     }
 
@@ -234,6 +431,23 @@ public class Lista {
         if (!(objeto instanceof Lista))
             return false;
         Lista lista = (Lista)objeto;
+        if (lista.getLongitud() != longitud) {
+            return false;
+        } else if (lista.getLongitud() == 0 && longitud == 0) {
+            return true;
+        } else {
+            Nodo explorador = cabeza;
+            Nodo gemelo = lista.cabeza;
+            while (explorador != null) {
+                if (explorador.elemento.equals(gemelo.elemento) == false) {
+                    return false;
+                }
+                explorador = explorador.siguiente;
+                gemelo = gemelo.siguiente;
+            }
+            return true;
+        }
+
         // Aquí va su código.
     }
 
@@ -242,6 +456,7 @@ public class Lista {
      * @return el nodo cabeza de la lista.
      */
     public Nodo getCabeza() {
+        return cabeza;
         // Aquí va su código.
     }
 
@@ -250,6 +465,7 @@ public class Lista {
      * @return el nodo rabo de la lista.
      */
     public Nodo getRabo() {
+        return rabo;
         // Aquí va su código.
     }
 }
