diff --git a/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java b/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
index 1e03d95..e24ffb0 100644
--- a/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
+++ b/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
@@ -4,6 +4,8 @@ import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.IOException;
 
+import mx.unam.ciencias.icc.Lista.Nodo;
+
 /**
  * Clase abstracta para bases de datos. Provee métodos para agregar y eliminar
  * registros, y para guardarse y cargarse de una entrada y salida dados. Además,
@@ -22,6 +24,7 @@ public abstract class BaseDeDatos {
      * Constructor único.
      */
     public BaseDeDatos() {
+        this.registros = new Lista();
         // Aquí va su código.
     }
 
@@ -30,6 +33,7 @@ public abstract class BaseDeDatos {
      * @return el número de registros en la base de datos.
      */
     public int getNumRegistros() {
+        return registros.getLongitud();
         // Aquí va su código.
     }
 
@@ -39,6 +43,7 @@ public abstract class BaseDeDatos {
      * @return una lista con los registros en la base de datos.
      */
     public Lista getRegistros() {
+        return registros.copia();
         // Aquí va su código.
     }
 
@@ -47,6 +52,7 @@ public abstract class BaseDeDatos {
      * @param registro el registro que hay que agregar a la base de datos.
      */
     public void agregaRegistro(Registro registro) {
+        registros.agregaFinal(registro);
         // Aquí va su código.
     }
 
@@ -55,6 +61,7 @@ public abstract class BaseDeDatos {
      * @param registro el registro que hay que eliminar de la base de datos.
      */
     public void eliminaRegistro(Registro registro) {
+        registros.elimina(registro);
         // Aquí va su código.
     }
 
@@ -62,6 +69,7 @@ public abstract class BaseDeDatos {
      * Limpia la base de datos.
      */
     public void limpia() {
+        registros.limpia();
         // Aquí va su código.
     }
 
@@ -71,6 +79,18 @@ public abstract class BaseDeDatos {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void guarda(BufferedWriter out) throws IOException {
+        Lista.Nodo explorador = registros.getCabeza();
+        while (explorador != null) {
+            try {
+                Object porGuardar = explorador.get();
+                Registro guardando = (Registro)porGuardar;
+                String textoGuardar = guardando.serializa();
+                out.write(textoGuardar);
+                explorador = explorador.getSiguiente();
+            } catch (IOException e) {
+                throw new IOException("ocurrio un error de salida");
+            }
+        }
         // Aquí va su código.
     }
 
@@ -82,6 +102,26 @@ public abstract class BaseDeDatos {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void carga(BufferedReader in) throws IOException {
+        if (!registros.esVacia()) {
+            registros.limpia(); 
+        }
+        String datos = in.readLine();
+            while (datos != null) {
+                datos = datos.strip();
+                if (datos.isEmpty()) {
+                    return;
+                }
+                if (datos != null) {
+                    Registro porAgregar = creaRegistro();
+                    try {
+                        porAgregar.deserializa(datos);
+                    } catch (ExcepcionLineaInvalida e) {
+                        throw new IOException("ocurrio un error de entrada");
+                    }
+                    registros.agregaFinal(porAgregar);
+                }
+                datos = in.readLine();
+            }
         // Aquí va su código.
     }
 
@@ -95,6 +135,18 @@ public abstract class BaseDeDatos {
      *         correcta.
      */
     public Lista buscaRegistros(Enum campo, Object valor) {
+        Lista regresar = new Lista();
+        Lista.Nodo explorador = registros.getCabeza();
+        while (explorador != null) {
+            if (explorador.get() instanceof Registro) {
+                Registro comparar = (Registro)explorador.get();
+                if (comparar.caza(campo, valor)) {
+                    regresar.agregaFinal(comparar);
+                }
+            }
+            explorador = explorador.getSiguiente();
+        }
+        return regresar;
         // Aquí va su código.
     }
 
diff --git a/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java b/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
index 921ac5b..1e4b766 100644
--- a/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
+++ b/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
@@ -10,6 +10,8 @@ public class BaseDeDatosEstudiantes extends BaseDeDatos {
      * @return un estudiante en blanco.
      */
     @Override public Registro creaRegistro() {
+        Estudiante regresar = new Estudiante(null, 0, 0.0, 0);
+        return regresar;
         // Aquí va su código.
     }
 }
diff --git a/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java b/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
index 4a8740d..fea783e 100644
--- a/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
+++ b/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
@@ -20,6 +20,18 @@ public enum CampoEstudiante {
      * @return una representación en cadena del campo.
      */
     @Override public String toString() {
+        switch (this) {
+            case NOMBRE:
+                return "Nombre";
+            case CUENTA:
+                return "# Cuenta";
+            case PROMEDIO:
+                return "Promedio";
+            case EDAD:
+                return "Edad";
+            default:
+                return null;
+        }
         // Aquí va su código.
     }
 }
diff --git a/src/main/java/mx/unam/ciencias/icc/Estudiante.java b/src/main/java/mx/unam/ciencias/icc/Estudiante.java
index 7027a7e..68bca10 100644
--- a/src/main/java/mx/unam/ciencias/icc/Estudiante.java
+++ b/src/main/java/mx/unam/ciencias/icc/Estudiante.java
@@ -29,6 +29,10 @@ public class Estudiante implements Registro {
                       int    cuenta,
                       double promedio,
                       int    edad) {
+        this.nombre    = nombre;
+        this.cuenta    = cuenta;
+        this.promedio  = promedio;
+        this.edad      = edad;
         // Aquí va su código.
     }
 
@@ -37,6 +41,7 @@ public class Estudiante implements Registro {
      * @return el nombre del estudiante.
      */
     public String getNombre() {
+        return nombre;
         // Aquí va su código.
     }
 
@@ -45,6 +50,7 @@ public class Estudiante implements Registro {
      * @param nombre el nuevo nombre del estudiante.
      */
     public void setNombre(String nombre) {
+        this.nombre = nombre;
         // Aquí va su código.
     }
 
@@ -53,6 +59,7 @@ public class Estudiante implements Registro {
      * @return el número de cuenta del estudiante.
      */
     public int getCuenta() {
+        return cuenta;
         // Aquí va su código.
     }
 
@@ -61,6 +68,7 @@ public class Estudiante implements Registro {
      * @param cuenta el nuevo número de cuenta del estudiante.
      */
     public void setCuenta(int cuenta) {
+        this.cuenta = cuenta;
         // Aquí va su código.
     }
 
@@ -69,6 +77,7 @@ public class Estudiante implements Registro {
      * @return el promedio del estudiante.
      */
     public double getPromedio() {
+        return promedio;
         // Aquí va su código.
     }
 
@@ -77,6 +86,7 @@ public class Estudiante implements Registro {
      * @param promedio el nuevo promedio del estudiante.
      */
     public void setPromedio(double promedio) {
+        this.promedio = promedio;
         // Aquí va su código.
     }
 
@@ -85,6 +95,7 @@ public class Estudiante implements Registro {
      * @return la edad del estudiante.
      */
     public int getEdad() {
+        return edad;
         // Aquí va su código.
     }
 
@@ -93,6 +104,7 @@ public class Estudiante implements Registro {
      * @param edad la nueva edad del estudiante.
      */
     public void setEdad(int edad) {
+        this.edad = edad;
         // Aquí va su código.
     }
 
@@ -101,6 +113,12 @@ public class Estudiante implements Registro {
      * @return una representación en cadena del estudiante.
      */
     @Override public String toString() {
+        String datitos = String.format("Nombre   : %s\n" +
+                                       "Cuenta   : %09d\n" +
+                                       "Promedio : %2.2f\n" +
+                                       "Edad     : %d",
+                                       nombre, cuenta, promedio, edad);
+	    return datitos;
         // Aquí va su código.
     }
 
@@ -113,6 +131,19 @@ public class Estudiante implements Registro {
      *         <code>false</code> en otro caso.
      */
     @Override public boolean equals(Object objeto) {
+        if (!(objeto instanceof Estudiante))
+            return false;
+        Estudiante estudiante = (Estudiante)objeto;
+        if ((this.nombre.equals(estudiante.nombre)) == false){
+            return false;
+        } else if ((this.cuenta == estudiante.cuenta) == false) {
+            return false;
+        } else if ((this.promedio == estudiante.promedio) == false){
+            return false;
+        } else if ((this.edad == estudiante.edad) == false){
+            return false;
+        }
+        return true;
         // Aquí va su código.
     }
 
@@ -123,6 +154,9 @@ public class Estudiante implements Registro {
      * @return la serialización del estudiante en una línea de texto.
      */
     @Override public String serializa() {
+        String regresar = String.format("%s\t%d\t%2.2f\t%d\n",
+                                     nombre, cuenta, promedio, edad);
+        return regresar;
         // Aquí va su código.
     }
 
@@ -135,6 +169,25 @@ public class Estudiante implements Registro {
      *         es una serialización válida de un estudiante.
      */
     @Override public void deserializa(String linea) {
+        if (linea == null) {
+            throw new ExcepcionLineaInvalida("la linea es nula");
+        }
+        if (linea.isEmpty()) {
+            throw new ExcepcionLineaInvalida("la linea esta vacía");
+        }
+        String porChecar = linea.strip();
+        String[] checar = porChecar.split("\t");
+        if (checar.length != 4){
+            throw new ExcepcionLineaInvalida("la linea no es un serialización válida");
+        }
+        this.nombre = checar[0];
+        try {
+            this.cuenta = Integer.parseInt(checar[1]);
+            this.promedio = Double.parseDouble(checar[2]);
+            this.edad = Integer.parseInt(checar[3]);
+        } catch (NumberFormatException e) {
+            throw new ExcepcionLineaInvalida("la linea no es un serialización válida");
+        }
         // Aquí va su código.
     }
 
@@ -145,9 +198,60 @@ public class Estudiante implements Registro {
      *         Estudiante}.
      */
     @Override public void actualiza(Registro registro) {
+        if (!(registro instanceof Estudiante)) {
+            throw new IllegalArgumentException("el registro no es instacia de Estudiante");
+        }
+        Estudiante nuevo = (Estudiante)registro;
+        this.nombre = nuevo.nombre;
+        this.cuenta = nuevo.cuenta;
+        this.promedio = nuevo.promedio;
+        this.edad = nuevo.edad;
         // Aquí va su código.
     }
 
+    /*
+    Metodos auxiliares para caza
+    */
+    private boolean cazaNombre(Object valor){
+        if (!(valor instanceof String)) {
+            return false;
+        }
+        if ((String)valor == "") {
+            return false;
+        }
+        if (nombre.contains((String)valor)) {
+            return true;
+        }
+        return false;
+    }
+    private boolean cazaCuenta(Object valor){
+        if (!(valor instanceof Integer)) {
+            return false;
+        }
+        if ((Integer)valor <= cuenta) {
+            return true;
+        }
+        return false;
+    }
+    private boolean cazaPromedio(Object valor){
+        if (!(valor instanceof Double)) {
+            return false;
+        }
+        if ((Double)valor <= promedio) {
+            return true;
+        }
+        return false;
+    }
+    private boolean cazaEdad(Object valor){
+        if (!(valor instanceof Integer)) {
+            return false;
+        }
+        if (((Integer)valor) <= edad) {
+            return true;
+        }
+        return false;
+    }
+
     /**
      * Nos dice si el estudiante caza el valor dado en el campo especificado.
      * @param campo el campo que hay que cazar.
@@ -175,6 +279,22 @@ public class Estudiante implements Registro {
      *         CampoEstudiante}.
      */
     @Override public boolean caza(Enum campo, Object valor) {
+        if (!(campo instanceof CampoEstudiante)) {
+            throw new IllegalArgumentException("el campo no es instancia de CampoEstudiante");
+        }
+        CampoEstudiante campito = (CampoEstudiante)campo;
+        switch (campito) {
+            case NOMBRE:
+                return cazaNombre(valor);
+            case CUENTA:
+                return cazaCuenta(valor);
+            case PROMEDIO:
+                return cazaPromedio(valor);
+            case EDAD:
+                return cazaEdad(valor);
+            default:
+                throw new IllegalArgumentException("el campo no es instancia de CampoEstudiante");
+        }
         // Aquí va su código.
     }
 }
diff --git a/src/main/java/mx/unam/ciencias/icc/Lista.java b/src/main/java/mx/unam/ciencias/icc/Lista.java
index 4502432..bbd40fc 100644
--- a/src/main/java/mx/unam/ciencias/icc/Lista.java
+++ b/src/main/java/mx/unam/ciencias/icc/Lista.java
@@ -28,6 +28,7 @@ public class Lista {
 
         /* Construye un nodo con un elemento. */
         private Nodo(Object elemento) {
+            this.elemento = elemento;
             // Aquí va su código.
         }
 
@@ -36,6 +37,7 @@ public class Lista {
          * @return el nodo anterior del nodo.
          */
         public Nodo getAnterior() {
+            return anterior;
             // Aquí va su código.
         }
 
@@ -44,6 +46,7 @@ public class Lista {
          * @return el nodo siguiente del nodo.
          */
         public Nodo getSiguiente() {
+            return siguiente;
             // Aquí va su código.
         }
 
@@ -52,6 +55,7 @@ public class Lista {
          * @return el elemento del nodo.
          */
         public Object get() {
+            return elemento;
             // Aquí va su código.
         }
     }
@@ -68,6 +72,7 @@ public class Lista {
      * @return la longitud de la lista, el número de elementos que contiene.
      */
     public int getLongitud() {
+        return longitud;
         // Aquí va su código.
     }
 
@@ -77,6 +82,7 @@ public class Lista {
      *         otro caso.
      */
     public boolean esVacia() {
+        return cabeza == null;
         // Aquí va su código.
     }
 
@@ -88,6 +94,18 @@ public class Lista {
      *         <code>null</code>.
      */
     public void agregaFinal(Object elemento) {
+        if (elemento == null) {
+            throw new IllegalArgumentException("el elemento es nulo");
+        }
+        Nodo nuevito = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevito;
+        } else {
+            rabo.siguiente = nuevito;
+            nuevito.anterior = rabo;
+            rabo = nuevito;
+        }
+        longitud++;
         // Aquí va su código.
     }
 
@@ -99,9 +117,85 @@ public class Lista {
      *         <code>null</code>.
      */
     public void agregaInicio(Object elemento) {
+        if (elemento == null) {
+            throw new IllegalArgumentException("el elemento es nulo");
+        }
+        Nodo nuevito = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevito;
+        } else {
+            cabeza.anterior = nuevito;
+            nuevito.siguiente = cabeza;
+            cabeza = nuevito;
+        }
+        longitud++;
         // Aquí va su código.
     }
 
+    /*
+    Metodo auxiliar para buscar el nodo donde esta un objeto (elemento)
+    Obtener Nodo dando objeto
+    */
+    private Nodo buscaNodo(Object buscado) {
+        if (buscado == null) {
+            return null;
+        }
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento.equals(buscado)) {
+                return explorador;
+            }
+            explorador = explorador.siguiente;
+        }
+        return null;
+    }
+
+    /*
+    Metodo auxiliar para buscar el nodo donde esta un objeto (elemento)
+    Obtener Nodo dando int
+    */
+    private Nodo buscaNodoConI(int i) {
+        if (i < 0 || i >= longitud) {
+            return null;
+        }
+        Nodo explorador = cabeza;
+        int contando = 0;
+        while (explorador != null) {
+            if (contando == i) {
+                return explorador;
+            }
+            explorador = explorador.siguiente;
+            contando++;
+        }
+        return null;
+    }
+
+    /*
+    Metodo auxiliar para eliminar un nodo
+    Eliminar Nodo dando el Nodo
+    */
+    private void desapareceNodo(Nodo condenado) {
+        if (condenado == null) {
+            return;
+        }
+        if (cabeza == rabo && cabeza == null) {
+            return;
+        }
+        longitud--;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else if (condenado == cabeza){
+            cabeza.siguiente.anterior = null;
+            cabeza = cabeza.siguiente;
+        } else if (condenado == rabo) {
+            rabo.anterior.siguiente = null;
+            rabo = rabo.anterior;
+        } else {
+            condenado.anterior.siguiente = condenado.siguiente;
+            condenado.siguiente.anterior = condenado.anterior;
+        }
+    }
+
     /**
      * Inserta un elemento en un índice explícito.
      *
@@ -118,6 +212,22 @@ public class Lista {
      *         <code>null</code>.
      */
     public void inserta(int i, Object elemento) {
+        if (elemento == null) {
+            throw new IllegalArgumentException("el elemento es nulo");
+        }
+        if (i <= 0) {
+            agregaInicio(elemento);
+        } else if (i >= longitud) {
+            agregaFinal(elemento);
+        } else {
+            Nodo intercambio = buscaNodoConI(i);
+            Nodo nuevo = new Nodo(elemento);
+            intercambio.anterior.siguiente = nuevo;
+            nuevo.anterior = intercambio.anterior;
+            intercambio.anterior = nuevo;
+            nuevo.siguiente = intercambio;
+            longitud++;
+        }
         // Aquí va su código.
     }
 
@@ -127,6 +237,14 @@ public class Lista {
      * @param elemento el elemento a eliminar.
      */
     public void elimina(Object elemento) {
+        if (elemento == null) {
+            return;
+        }
+        Nodo marcado = buscaNodo(elemento);
+        if (marcado == null) {
+            return;
+        }
+        desapareceNodo(marcado);
         // Aquí va su código.
     }
 
@@ -136,6 +254,12 @@ public class Lista {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public Object eliminaPrimero() {
+        if (esVacia()) {
+            throw new NoSuchElementException("la lista es vacía");
+        }
+        Object porEliminar = cabeza.elemento;
+        desapareceNodo(cabeza);
+        return porEliminar;
         // Aquí va su código.
     }
 
@@ -145,6 +269,12 @@ public class Lista {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public Object eliminaUltimo() {
+        if (esVacia()) {
+            throw new NoSuchElementException("la lista es vacía");
+        }
+        Object porEliminar = rabo.elemento;
+        desapareceNodo(rabo);
+        return porEliminar;
         // Aquí va su código.
     }
 
@@ -155,6 +285,7 @@ public class Lista {
      *         <code>false</code> en otro caso.
      */
     public boolean contiene(Object elemento) {
+        return buscaNodo(elemento) != null;
         // Aquí va su código.
     }
 
@@ -163,6 +294,13 @@ public class Lista {
      * @return una nueva lista que es la reversa la que manda llamar el método.
      */
     public Lista reversa() {
+        Lista clonada = new Lista();
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            clonada.agregaInicio(explorador.elemento);
+            explorador = explorador.siguiente;
+        }
+        return clonada;
         // Aquí va su código.
     }
 
@@ -172,6 +310,13 @@ public class Lista {
      * @return una copiad de la lista.
      */
     public Lista copia() {
+        Lista clonada = new Lista();
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            clonada.agregaFinal(explorador.elemento);
+            explorador = explorador.siguiente;
+        }
+        return clonada;
         // Aquí va su código.
     }
 
@@ -179,6 +324,8 @@ public class Lista {
      * Limpia la lista de elementos, dejándola vacía.
      */
     public void limpia() {
+        cabeza = rabo = null;
+        longitud = 0;
         // Aquí va su código.
     }
 
@@ -188,6 +335,11 @@ public class Lista {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public Object getPrimero() {
+        if (esVacia()) {
+            throw new NoSuchElementException("la lista es vacía");
+        } else {
+            return cabeza.elemento;
+        }
         // Aquí va su código.
     }
 
@@ -197,6 +349,11 @@ public class Lista {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public Object getUltimo() {
+        if (esVacia()) {
+            throw new NoSuchElementException("la lista es vacía");
+        } else {
+            return rabo.elemento;
+        }
         // Aquí va su código.
     }
 
@@ -208,6 +365,19 @@ public class Lista {
      *         igual que el número de elementos en la lista.
      */
     public Object get(int i) {
+        if (i < 0 || i >= longitud) {
+            throw new ExcepcionIndiceInvalido("el indice no esta en el rango de la lista");
+        }
+        Nodo explorador = cabeza;
+        int contando = 0;
+        while (explorador != null) {
+            if (contando == i) {
+                return explorador.elemento;
+            }
+            explorador = explorador.siguiente;
+            contando++;
+        }
+        return null;
         // Aquí va su código.
     }
 
@@ -218,6 +388,16 @@ public class Lista {
      *         no está contenido en la lista.
      */
     public int indiceDe(Object elemento) {
+        int medidor = 0;
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento.equals(elemento)) {
+                return medidor;
+            }
+            explorador = explorador.siguiente;
+            medidor++;
+        }
+        return -1;
         // Aquí va su código.
     }
 
@@ -226,6 +406,23 @@ public class Lista {
      * @return una representación en cadena de la lista.
      */
     @Override public String toString() {
+        if (esVacia()) {
+            String cadenitaRegreso = "[]";
+            return cadenitaRegreso;
+        }
+        String cadenitaRegreso = "[";
+        Nodo explorador = cabeza;
+        while (explorador != null) {
+            if (explorador.elemento != null) {
+                cadenitaRegreso += explorador.elemento;
+            }
+            if (explorador.siguiente != null) {
+                cadenitaRegreso += ", ";
+            }
+            explorador = explorador.siguiente;
+        }
+        cadenitaRegreso += "]";
+        return cadenitaRegreso;
         // Aquí va su código.
     }
 
@@ -236,6 +433,25 @@ public class Lista {
      *         <code>false</code> en otro caso.
      */
     @Override public boolean equals(Object objeto) {
+        if (!(objeto instanceof Lista))
+            return false;
+        Lista lista = (Lista)objeto;
+        if (lista.getLongitud() != longitud) {
+            return false;
+        } else if (lista.getLongitud() == 0 && longitud == 0) {
+            return true;
+        } else {
+            Nodo explorador = cabeza;
+            Nodo gemelo = lista.cabeza;
+            while (explorador != null) {
+                if (explorador.elemento.equals(gemelo.elemento) == false) {
+                    return false;
+                }
+                explorador = explorador.siguiente;
+                gemelo = gemelo.siguiente;
+            }
+            return true;
+        }
         // Aquí va su código.
     }
 
@@ -244,6 +460,7 @@ public class Lista {
      * @return el nodo cabeza de la lista.
      */
     public Nodo getCabeza() {
+        return cabeza;
         // Aquí va su código.
     }
 
@@ -252,6 +469,7 @@ public class Lista {
      * @return el nodo rabo de la lista.
      */
     public Nodo getRabo() {
+        return rabo;
         // Aquí va su código.
     }
 }
